Enhance filters description in DataCollector/WebProfiler

antograssiot opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

before:
<img width="857" alt="capture d ecran 2018-04-13 a 21 03 50" src="https://user-images.githubusercontent.com/4977112/38753362-63a06528-3f5e-11e8-9c00-95680eb7cadb.png">

after:
<img width="859" alt="capture d ecran 2018-04-13 a 21 01 00" src="https://user-images.githubusercontent.com/4977112/38753371-6ba2de7c-3f5e-11e8-9e6f-8c874caa7c4a.png">

and a warning in the toolbar:
<img width="232" alt="capture d ecran 2018-04-13 a 21 00 33" src="https://user-images.githubusercontent.com/4977112/38753391-7b4f8f14-3f5e-11e8-8857-6cda0084d780.png">
<img width="216" alt="capture d ecran 2018-04-13 a 21 00 48" src="https://user-images.githubusercontent.com/4977112/38753394-83cdfdce-3f5e-11e8-997d-2ccf193fa93b.png">

or when no filters are available:
<img width="856" alt="capture d ecran 2018-04-14 a 08 36 46" src="https://user-images.githubusercontent.com/4977112/38765299-18bfa472-3fbf-11e8-89eb-3b6625e159c3.png">

antograssiot commented:
To better understand, the whole page would be :
![screenshot-2018-4-13 symfony profiler](https://user-images.githubusercontent.com/4977112/38753559-063eed54-3f5f-11e8-9bc1-cba891de63cc.png)

IonBazan commented on src/Bridge/Symfony/Bundle/Resources/views/DataCollector/request.html.twig:
```diff
@@ -86,22 +93,26 @@
+                    <tr{% if filter is same as(null) %} class="status-warning"{% endif %}>
+                        <td>
+                            {{ id }}
+                            {% if filter is same as(null) %}<span class="newline text-muted">ignored filter</span>{% else %}{{ dump(filter) }}{% endif %}
```
This condition is copy-pasted above. You could bind it to a variable.
Also, make it multiline.

antograssiot replied:
thanks @IonBazan, it has been updated

antograssiot commented:
Also added a message when there's no collection operation.
<img width="862" alt="capture d ecran 2018-04-14 a 15 50 19" src="https://user-images.githubusercontent.com/4977112/38768978-c216bf38-3ffb-11e8-89dd-59a910f27fb8.png">

antograssiot commented on src/Bridge/Symfony/Bundle/Resources/views/DataCollector/request.html.twig:
```diff
@@ -60,6 +67,12 @@
             <tbody>
                 {% for key, itemOperation in collector.resourceMetadata.itemOperations %}
                     {{ apiPlatform.operationLine(key, itemOperation, collector.requestAttributes.item_operation_name|default('')) }}
+                {% else %}
```
maybe this block is not needed as there is always at least one item operation so the IRI can be generated, is this assumption correct ?

dunglas replied:
Yes, you can drop it.

dunglas commented on src/Bridge/Symfony/Bundle/DataCollector/RequestDataCollector.php:
```diff
@@ -25,25 +26,44 @@
      */
     public function collect(Request $request, Response $response, \Exception $exception = null)
     {
+        $counters = [
```
Can be inlined.

dunglas commented on src/Bridge/Symfony/Bundle/DataCollector/RequestDataCollector.php:
```diff
@@ -25,25 +26,44 @@
         $resourceMetadata = $resourceClass ? $this->metadataFactory->create($resourceClass) : null;
 
+        $filters = [];
+        foreach (($resourceMetadata ? $resourceMetadata->getAttribute('filters', []) : []) as $id) {
```
The parenthesis are useless, right?

dunglas commented on src/Bridge/Symfony/Bundle/Resources/views/DataCollector/request.html.twig:
```diff
@@ -11,6 +11,7 @@
 
 {% block toolbar %}
     {% set icon %}
+        {% set status_color = collector.counters.ignored_filters|default(0) ? 'yellow': 'default' %}
```
Space to add between `yellow` and `:`
Can also be `default(false)`

dunglas commented on src/Bridge/Symfony/Bundle/Resources/views/DataCollector/request.html.twig:
```diff
@@ -19,9 +20,15 @@
             <b>Resource Class</b>
             <span>{{ collector.resourceClass|default('Not an API Platform resource') }}</span>
         </div>
+        {% if collector.counters.ignored_filters|default(0) %}
```
Same here `default(false)`

dunglas commented on tests/Bridge/Symfony/Bundle/DataCollector/RequestDataCollectorTest.php:
```diff
@@ -36,6 +37,8 @@ class RequestDataCollectorTest extends TestCase
 
     private $metadataFactory;
 
+    private $filterLocator;
```
Usually, we don't add a blank line between props when there are no doc block.

Simperfit commented:
Is this going to fix #102 ?

antograssiot commented:
Updated and rebased @dunglas.
@Simperfit it only address part of it as for now you'll get filters enabled for a given resource, a list of the ignored/badly declared filtres but not the list of the active filtres on the request for exemple.

dunglas commented:
Thanks @antograssiot!
